newton/geometry/inertia.py

```diff
@@ -485,3 +487,266 @@ def compute_shape_inertia(
+        was_corrected = True
+        # Fallback: use diagonal elements
+        trace = np.trace(corrected_inertia)
+        if trace <= 0:
+            trace = 1e-6
+        corrected_inertia = np.eye(3) * (trace / 3.0)
+        has_violations = False
+        principal_moments = [trace / 3.0, trace / 3.0, trace / 3.0]
+
+    if has_violations:
+        warnings.warn(
+            f"Inertia tensor{body_id} violates triangle inequality with principal moments ({I1:.6f}, {I2:.6f}, {I3:.6f})",
+            stacklevel=2,
+        )
+
+        if balance_inertia:
+            # For non-diagonal matrices, we need to adjust while preserving the rotation
+            deficit = I3 - I1 - I2
+            if deficit > 0:
+                # Simple approach: add scalar to all eigenvalues to ensure validity
+                # This preserves eigenvectors exactly
+                # We need: (I1 + a) + (I2 + a) >= I3 + a
+                # Which simplifies to: I1 + I2 + a >= I3
+                # So: a >= I3 - I1 - I2 = deficit
+                adjustment = deficit + 1e-6
+
+                # Add scalar*I to shift all eigenvalues equally
+                corrected_inertia = corrected_inertia + np.eye(3) * adjustment
+                was_corrected = True
+
+                # Update principal moments
+                new_I1 = I1 + adjustment
+                new_I2 = I2 + adjustment
+                new_I3 = I3 + adjustment
+
+                warnings.warn(
+                    f"Balanced principal moments{body_id} from ({I1:.6f}, {I2:.6f}, {I3:.6f}) to "
+                    f"({new_I1:.6f}, {new_I2:.6f}, {new_I3:.6f})",
+                    stacklevel=2,
+                )
+
+    # Final check: ensure the corrected inertia matrix is positive definite
+    if has_violations and balance_inertia:
+        # Need to recompute after balancing since we modified the matrix
+        try:
+            eigenvalues = np.linalg.eigvals(corrected_inertia)
+        except np.linalg.LinAlgError:
+            warnings.warn(f"Failed to compute eigenvalues of inertia matrix{body_id}", stacklevel=2)
+            eigenvalues = np.array([0.0, 0.0, 0.0])
+
+    # Check final eigenvalues
+    if np.any(eigenvalues <= 0) or np.any(~np.isfinite(eigenvalues)):
+        warnings.warn(
+            f"Corrected inertia matrix{body_id} is not positive definite, this should not happen", stacklevel=2
+        )
+        # As a last resort, make it positive definite by adding a small value to diagonal
+        min_eigenvalue = np.min(eigenvalues[np.isfinite(eigenvalues)]) if np.any(np.isfinite(eigenvalues)) else -1e-6
+        epsilon = abs(min_eigenvalue) + 1e-6
+        corrected_inertia[0, 0] += epsilon
+        corrected_inertia[1, 1] += epsilon
+        corrected_inertia[2, 2] += epsilon
+        was_corrected = True
+
+    return corrected_mass, wp.mat33(corrected_inertia), was_corrected
+
+
+@wp.kernel
```

Comment:
We could improve the compilation time here by using 
```py
@wp.kernel(enable_backward=False, module="unique")
```
see https://github.com/newton-physics/newton/blob/main/newton/geometry/kernels.py#L770